From: Linus Torvalds <torvalds@linux-foundation.org>
Date: Sat, 21 Feb 2026 15:12:09 -0800
Subject: slab.h: disable completely broken overflow handling in flex
 allocations
Git-commit: fa5c82f4d2bbde10e9fd3a32aecacfe3813919ba
Patch-mainline: v7.0-rc1
References: bsc#1280139

Commit 69050f8d6d07 ("treewide: Replace kmalloc with kmalloc_obj for
non-scalar types") started using the new allocation helpers, and in the
process showed that they were completely non-working.

The overflow logic in overflows_flex_counter_type() is completely the
wrong way around, and that broke __alloc_flex() completely.  By chance,
the resulting code was then such a mess that clang generated
sufficiently garbage code that objtool warned about it all.  Which made
it somewhat quicker to narrow things down.

While fixing overflows_flex_counter_type() would presumably fix this
all, I'm excising the whole broken overflow logic from __alloc_flex(),
because we don't want that kind of code in basic allocation functions
anyway.

That (no longer) broken overflows_flex_counter_type() thing needs to be
inserted into the actual __set_flex_counter() logic in the unlikely case
that we ever want this at all.  And made conditional.

Fixes: 81cee9166a90 ("compiler_types: Introduce __flex_counter() and family")
Fixes: 69050f8d6d07 ("treewide: Replace kmalloc with kmalloc_obj for non-scalar types")
Cc: Kees Cook <kees@kernel.org>
Link: https://lore.kernel.org/all/CAHk-=whEd020BYzGTzYrENjD9Z5_82xx6h8HsQvH5xDSnv0=Hw@mail.gmail.com/
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Vlastimil Babka <vbabka@suse.cz>
---
 include/linux/overflow.h |    2 +-
 include/linux/slab.h     |    6 +-----
 2 files changed, 2 insertions(+), 6 deletions(-)

--- a/include/linux/overflow.h
+++ b/include/linux/overflow.h
@@ -477,7 +477,7 @@ static inline size_t __must_check size_s
  * @FAM is not annotated with __counted_by(), always returns true.
  */
 #define overflows_flex_counter_type(TYPE, FAM, COUNT)		\
-	(!overflows_type(COUNT, typeof_flex_counter(((TYPE *)NULL)->FAM)))
+	(overflows_type(COUNT, typeof_flex_counter(((TYPE *)NULL)->FAM)))
 
 /**
  * __set_flex_counter() - Set the counter associated with the given flexible
--- a/include/linux/slab.h
+++ b/include/linux/slab.h
@@ -920,11 +920,7 @@ static __always_inline __alloc_size(1) v
 ({									\
 	const size_t __count = (COUNT);					\
 	const size_t __obj_size = struct_size_t(TYPE, FAM, __count);	\
-	TYPE *__obj_ptr;						\
-	if (WARN_ON_ONCE(overflows_flex_counter_type(TYPE, FAM,	__count))) \
-		__obj_ptr = NULL;					\
-	else								\
-		__obj_ptr = KMALLOC(__obj_size, GFP);			\
+	TYPE *__obj_ptr = KMALLOC(__obj_size, GFP);			\
 	if (__obj_ptr)							\
 		__set_flex_counter(__obj_ptr->FAM, __count);		\
 	__obj_ptr;							\
